src/java.base/share/classes/javax/security/auth/Subject.java

```diff
@@ -340,23 +385,30 @@ public Subject run() {
      */
     @SuppressWarnings("removal")
     public static Subject current() {
-        return getSubject(AccessController.getContext());
+        if (!SharedSecrets.getJavaLangAccess().allowSecurityManager()) {
+            return SCOPED_SUBJECT.orElse(null);
+        } else {
+            return getSubject(AccessController.getContext());
+        }
     }
 
     /**
      * Executes a {@code Callable} with {@code subject} as the
      * current subject.
      *
-     * @implNote
-     * This method calls {@link #doAs(Subject, PrivilegedExceptionAction)
-     * Subject.doAs(subject, altAction)} which stores the subject in
-     * a new {@code AccessControlContext}, where {@code altAction.run()}
-     * is equivalent to {@code action.call()} and the exception thrown is
-     * modified to match the specification of this method. This preserves
-     * compatibility with code that may still be calling
-     * {@code getSubject(AccessControlContext)} which retrieves the subject
-     * from an {@code AccessControlContext}. This behavior is subject
-     * to change in a future version.
+     * <p> If a security manager is <a href=#sm-not-allowed>not allowed</a>,
+     * this method launches {@code action} and binds {@code subject} to the
+     * period of its execution.
+     * <p> Otherwise, this method first retrieves the
+     * current Thread's {@code AccessControlContext} via
+     * {@code AccessController.getContext},
+     * and then instantiates a new {@code AccessControlContext}
+     * using the retrieved context along with a new
+     * {@code SubjectDomainCombiner} (constructed using
+     * the provided {@code Subject}).
+     * Finally, this method invokes {@code AccessController.doPrivileged},
+     * passing it the provided {@code PrivilegedAction},
+     * as well as the newly constructed {@code AccessControlContext}.
```

Comment:
I think this method would be easier to present if the allowed and not-allowed cases were in separate paragraphs. The reason is that the SM allowed case has a lot more test. For the not allowed case then it would be clearer to say that the calls the value returning function with the current Thread bound to the given Subject.